chichiwang/exercise-js
# Add scripts: `build:prod`, `build:prod:analyze`, `build:prod:run`

## Description
Linked to Issue: #45 
Configure a production build script that minifies JS/CSS assets (and actually builds CSS assets).

TODOs have been created in Issue #45 for further optimizations to the production build process.

## Changes
* Scripts added to `package.json`:
  * `build:prod`: Build minified production assets to `/dist`
  * `build:prod:analyze`: Build minified production assets to `/dist` and then run bundle analyzer on them
  * `build:prod:run`: Build minified production assets to `/dist` and then run the built server asset
* DevDependency `mini-css-extract-plugin` added to `package.json`
* Create a new webpack config utility `webpack/env.js`
  * Determine if `NODE_ENV` env variable is set to production, otherwise consider it development
* Create a new webpack config utility `webpack/sassRules.js`
  * Contains all of the SASS module rules for webpack for both the client and server
* Update `webpack/base.js` config:
  * Set `mode` configuration to the detected `NODE_ENV` env variable (development/production)
    * This will minify assets for production
  * Set `devtool` source-mapping configuration based on if mode is production or not
    * Use `source-map` for production and `eval-source-map` for non-production builds
* Update config in `webpack/client.js`
  * Configured to use client SASS rules and `MiniCssExtractPlugin`
* Update config in `webpack/server.js`
  * Configured to use server SASS rules
  * Configured to use `NodemonPlugin` only in non-production builds
* Add a static assets route `/styles` to Express server in `server/index.ts`
* Add a `<Link />` tag to the head of the document in `app/App.tsx` to load in `/styles/app.css` file

## Steps to QA
* Pull down and switch to this branch
* Run `npm run build:dev:run` to ensure assets are built correctly in `dist/`
  * Verify behavior in browser
* Run `npm run build:dev:analyze` to observe the bundles created and their bundle sizes

![image](https://github.com/chichiwang/exercise-js/assets/2304118/bf518da1-abda-4bb5-aa4a-297408717d6a)

* Run `npm run build:prod:run` to ensure assets are built correctly in `dist/`
  * Verify behavior in browser
* Run `npm run build:prod:analyze` to observe the bundles are created and their bundle sizes

![image](https://github.com/chichiwang/exercise-js/assets/2304118/7333cc7d-6d1c-43b3-9398-301271c730c1)

  * Ensure the bundle sizes are smaller than the dev build counterparts, ensuring minification
